src/dir-hamt-sharded.js

```diff
@@ -27,41 +32,180 @@ function shardedDirExporter (cid, node, name, path, pathRest, resolve, size, dag
+        resolve
+      )
+    ]
+
+    // place dir before if not specifying subtree
     streams.unshift(values([dir]))
+
+    return cat(streams)
+  }
+
+  const deferred = defer.source()
+  const targetFile = pathRest[0]
+
+  // recreate our level of the HAMT so we can load only the subshard in pathRest
+  waterfall([
+    (cb) => {
+      if (!options.rootBucket) {
+        options.rootBucket = new Bucket({
+          hashFn: DirSharded.hashFn
+        })
+        options.hamtDepth = 1
+
+        return addLinksToHamtBucket(node.links, options.rootBucket, options.rootBucket, cb)
+      }
+
+      return addLinksToHamtBucket(node.links, options.lastBucket, options.rootBucket, cb)
+    },
+    (cb) => findPosition(targetFile, options.rootBucket, cb),
+    ({ position }, cb) => {
+      let prefix = toPrefix(position.pos)
+      const bucketPath = toBucketPath(position)
+
+      if (bucketPath.length > (options.hamtDepth)) {
+        options.lastBucket = bucketPath[options.hamtDepth]
+
+        prefix = toPrefix(options.lastBucket._posAtParent)
+      }
+
+      const streams = [
+        pull(
+          values(node.links),
+          map((link) => {
+            const entryPrefix = link.name.substring(0, 2)
+            const entryName = link.name.substring(2)
+            const entryPath = entryName ? path + '/' + entryName : path
+
+            if (entryPrefix !== prefix) {
+              // not the entry or subshard we're looking for
+              return false
+            }
+
+            if (entryName && entryName !== targetFile) {
+              // not the entry we're looking for
+              return false
+            }
+
+            if (!entryName) {
+              // we are doing to descend into a subshard
+              options.hamtDepth++
+            } else {
+              // we've found the node we are looking for, remove the context
+              // so we don't affect further hamt traversals
+              delete options.rootBucket
+              delete options.lastBucket
+              delete options.hamtDepth
+            }
+
+            return {
+              depth: entryName ? depth + 1 : depth,
+              name: entryName,
+              path: entryPath,
+              multihash: link.cid.buffer,
+              pathRest: entryName ? pathRest.slice(1) : pathRest,
+              parent: dir || parent
+            }
+          }),
+          filter(Boolean),
+          resolve
+        )
+      ]
+
+      if (options.fullPath) {
+        streams.unshift(values([dir]))
+      }
+
+      cb(null, streams)
+    }
+  ], (err, streams) => {
+    if (err) {
+      return deferred.resolve(error(err))
+    }
+
+    deferred.resolve(cat(streams))
+  })
+
+  return deferred
+}
+
+const addLinksToHamtBucket = (links, bucket, rootBucket, callback) => {
+  Promise.all(
+    links.map(link => {
+      if (link.name.length === 2) {
+        const pos = parseInt(link.name, 16)
+
+        return bucket._putObjectAt(pos, new Bucket({
```

Comment:
I know, but we control that module.  I can make them all public, but they aren't really of use unless you're trying to predict shard entry indices.

Reply:
TBH all of this code used to be in the same module but we decomposed it.  It's all really tightly coupled still, could definitely be improved in future.

Reply:
Understood, but the exact problem I want to avoid happened with ipfs and dag-pb when the `_multihash` property was removed - we control both of those modules too!

If you want to deal with it in a separate PR that's cool, but lets at least make sure it does finally get done, maybe raise an issue?